pulsar-client-cpp/lib/ClientConnection.cc

```diff
@@ -108,7 +109,52 @@ ClientConnection::ClientConnection(const std::string& endpoint, ExecutorServiceP
           outgoingBuffer_(SharedBuffer::allocate(DefaultBufferSize)),
           outgoingCmd_(),
           havePendingPingRequest_(false),
-          keepAliveTimer_() {
+          keepAliveTimer_(),
+          isTlsAllowInsecureConnection_(false) {
+    if (clientConfiguration.isUseTls()) {
+        using namespace boost::filesystem;
+
+        boost::asio::ssl::context ctx(boost::asio::ssl::context::tlsv12_client);
+
+        if (clientConfiguration.isTlsAllowInsecureConnection()) {
+            ctx.set_verify_mode(boost::asio::ssl::verify_none);
+            isTlsAllowInsecureConnection_ = true;
+        } else {
+            ctx.set_verify_mode(boost::asio::ssl::verify_peer);
+            std::string trustCertFilePath = clientConfiguration.getTlsTrustCertsFilePath();
+            if (exists(path(trustCertFilePath))) {
+                ctx.load_verify_file(trustCertFilePath);
+            } else {
+                LOG_ERROR(trustCertFilePath << ": No such trustCertFile");
+                close();
+                return;
+            }
+        }
+
+        AuthenticationDataPtr authData;
+        if (authentication_->getAuthData(authData) == ResultOk && authData->hasDataForTls()){
+            std::string tlsCertificates = authData->getTlsCertificates();
+            std::string tlsPrivateKey = authData->getTlsPrivateKey();
+
+            if (exists(path(tlsCertificates))) {
+                ctx.use_certificate_file(tlsCertificates, boost::asio::ssl::context::pem);
+            } else {
+                LOG_ERROR(tlsCertificates << ": No such tlsCertificates");
+                close();
+                return;
+            }
+
+            if (exists(path(tlsPrivateKey))) {
+                ctx.use_private_key_file(tlsPrivateKey, boost::asio::ssl::context::pem);
+            } else {
+                LOG_ERROR(tlsPrivateKey << ": No such tlsPrivateKey");
+                close();
+                return;
+            }
+        }
+
+        tlsSocket_ = executor->createTlsSocket(socket_, ctx);
+    }
```

Comment:
I think we would have better initialize socket_ at line 101 because if we don't do it, createTlsSocket doesn't work with a following error.
```
Assertion failed: (px != 0), function operator*, file /usr/local/include/boost/smart_ptr/shared_ptr.hpp, line 703.
```